data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/SqsDynamicVisibilityTimeoutManager.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.source;
+
+import org.opensearch.dataprepper.plugins.source.configuration.SqsOptions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.sqs.SqsClient;
+import software.amazon.awssdk.services.sqs.model.ChangeMessageVisibilityRequest;
+import software.amazon.awssdk.services.sqs.model.Message;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+public class SqsDynamicVisibilityTimeoutManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SqsDynamicVisibilityTimeoutManager.class);
+    static final int VISIBILITY_TIMEOUT_SECONDS_PER_QUARTER_GB = 30;
+    static final int QUARTER_GB_BYTES = 250_000_000;
+    static final int MAX_VISIBILITY_TIMEOUT_SECONDS = 43_200;
+    static final int MINIMUM_WAIT_BEFORE_CHANGE_VISIBILITY_TIMEOUT_SECONDS = 2;
+    static final int TIME_BETWEEN_CHANGE_MESSAGE_CALL_AND_VISIBILITY_TIMEOUT_SECONDS = 10;
+
+    private final SqsClient sqsClient;
+    private final ScheduledExecutorService scheduledExecutorService;
+    private final String queueUrl;
+    private final int initialMessageVisibilityTimeout;
+    private final int initialWaitBeforeIncrease;
+
+    private final Map<String, ScheduledFuture<?>> messageToScheduledFuture;
+
+    public SqsDynamicVisibilityTimeoutManager(final SqsClient sqsClient,
+                                              final SqsOptions sqsOptions) {
+        this.sqsClient = sqsClient;
+        this.scheduledExecutorService = Executors.newScheduledThreadPool(sqsOptions.getMaximumMessages());
```

Comment:
I'd like to see if we can remove this thread pool altogether. But, maybe that isn't quite possible.

If so, please customize this with two changes:
1. Name the threads to help with debugging.
2. Make the threads daemon threads since they are not doing core processing.

See `AcknowledgementSetMonitorThread` or #2483 where I added `AcknowledgementSetMonitorThread` for how to do this.